--- src/Makefile.orig	2010-05-01 19:23:29.000000000 +0200
+++ src/Makefile	2010-05-01 19:23:48.000000000 +0200
@@ -61,7 +61,7 @@
 # Compiler options
 ifeq ($(GSPICEUI_DBG),0)
   # Options for release (not using -Wall since it's GCC specific)
-  CXXFLAGS := -O -pipe $(shell $(WXCFG) --cxxflags)
+  CXXFLAGS := -O -pipe $(shell $(WXCFG) --cxxflags) %%CXX_ARCHFLAGS%%
 else
   # Options for development
   CXXFLAGS := -Wall -g -pipe $(shell $(WXCFG) --cxxflags)
@@ -72,7 +72,7 @@
 
 # Libraries
 # (The pkg-config stuff was requested by a user, somehow pangox was missing)
-LIBS := $(shell $(WXCFG) --libs core,base,html) $(shell pkg-config --libs pangox)
+LIBS := $(shell $(WXCFG) --libs core,base,html) $(shell pkg-config --libs pangox) -lstdc++
 
 # Objects
 OBJS := $(wildcard *.cpp) $(wildcard */*.cpp) $(wildcard */*/*.cpp)
