Index: files/patch-configure-change_install_names.diff
===================================================================
--- files/patch-configure-change_install_names.diff	(revision 86008)
+++ files/patch-configure-change_install_names.diff	(working copy)
@@ -1,5 +1,18 @@
---- configure.orig	2011-01-13 02:34:25.000000000 -0800
-+++ configure	2011-01-13 02:33:45.000000000 -0800
+diff -x .svn -ru orig/configure patch/configure
+--- configure	2011-10-19 17:36:57.000000000 +0200
++++ configure	2011-10-19 18:01:41.000000000 +0200
+@@ -18888,9 +18888,9 @@
+ echo "$as_me: WARNING: Please use --with-macosx-sdk=PATH and --enable-universal_binary without an argument" >&2;}
+             fi
+         fi
+-                OSX_ARCH_OPTS="-arch ppc -arch i386"
++                OSX_ARCH_OPTS="@@MP_ARCH_FLAGS@@"
+ 	if test "$wxUSE_OSX_COCOA" = 1; then
+-            OSX_ARCH_OPTS="$OSX_ARCH_OPTS -arch x86_64"
++            OSX_ARCH_OPTS="$OSX_ARCH_OPTS"
+         fi
+         { echo "$as_me:$LINENO: checking for universal binary architectures" >&5
+ echo $ECHO_N "checking for universal binary architectures... $ECHO_C" >&6; }
 @@ -35366,16 +35366,18 @@
              DYLIB_RPATH_POSTLINK="${HOST_PREFIX}install_name_tool -id \$@ \$@"
              cat <<EOF >change-install-names
@@ -14,10 +27,10 @@
 +changes=''
 +for dep in \$libnames; do
 +    changes="\$changes -change \$4/\$dep \$3/\$dep"
- done
++done
 +for i in \$libnames; do
 +    ${HOST_PREFIX}install_name_tool \$changes -id \$3/\$i \$1/\$i
-+done
+ done
 +${HOST_PREFIX}install_name_tool \$changes \$2/wxrc-2.9
  EOF
              chmod +x change-install-names
@@ -26,8 +39,9 @@
          fi
  
                                          HEADER_PAD_OPTION="-headerpad_max_install_names"
---- configure.in.orig	2011-01-13 02:34:15.000000000 -0800
-+++ configure.in	2011-01-13 02:32:44.000000000 -0800
+diff -x .svn -ru orig/configure.in patch/configure.in
+--- configure.in	2011-10-19 17:36:57.000000000 +0200
++++ configure.in	2011-10-19 17:38:38.000000000 +0200
 @@ -4004,16 +4004,18 @@
              DYLIB_RPATH_POSTLINK="${HOST_PREFIX}install_name_tool -id \$@ \$@"
              cat <<EOF >change-install-names
Index: Portfile
===================================================================
--- Portfile	(revision 86008)
+++ Portfile	(working copy)
@@ -7,7 +7,7 @@
 conflicts       wxgtk wxWidgets
 version         2.9.2
 revision        1
-license         wxWindow Licence 3.1
+license         wxwidgets-3.1
 categories      graphics devel
 platforms       darwin
 maintainers     jwa
@@ -55,6 +55,10 @@
 patchfiles  patch-configure-change_install_names.diff
 configure.cmd   ../configure
 configure.ldflags   -L${build.dir}/lib -L${prefix}/lib
+
+configure.env-append CPP=${developer_dir}/usr/bin/llvm-cpp-4.2
+configure.env-append CXXCPP=${developer_dir}/usr/bin/llvm-cpp-4.2
+
 configure.args  --mandir=${prefix}/share/man \
                 --with-libiconv-prefix=${prefix} \
                 --with-libjpeg \
@@ -70,12 +74,20 @@
 
 build.target
 
-universal_variant   no
+universal_variant   yes
 use_parallel_build  yes
 
+
 configure.ccache    no
-configure.cppflags-append -arch ${configure.build_arch}
 
+if {![variant_isset universal]} {
+    set mp_arch_flags   "-arch ${configure.build_arch}"
+} else {
+    set mp_arch_flags   "-arch [join ${universal_archs} \ -arch\ ]"
+}
+   
+configure.cppflags-append "${mp_arch_flags}"
+
 post-destroot {
     set confscript ${prefix}/lib/wx/config/osx_cocoa-unicode-2.9
     ln -sf ${confscript} ${destroot}${prefix}/bin/wx-config
@@ -85,25 +97,14 @@
      configure.args-append \
      --with-macosx-sdk=/Developer/SDKs/MacOSX10.6.sdk \
      --with-macosx-version-min=10.6
+}
 
-    universal_variant yes
-    variant universal {
+variant universal {
         configure.args-append   --enable-universal_binary
+}
 
-        #removing single arch target
-        configure.cppflags-delete -arch ${configure.build_arch}
-        
-        #enabling the following breaks a cpp test while configuring
-        #it's not needed as archs are hardcoded in configure script
-#        configure.cppflags-append "-arch [join ${universal_archs} \ -arch\ ]"
-
-        #removing hardcoded ppc target as it won't build with XCode 4.1
-        #this is required due the use of --enable-universal_binary
-        #it could stay out of universal variant as ppc platform isn't supported in darwin 11
-        post-patch {
-            reinplace -- "s|-arch ppc ||" ${worksrcpath}/../configure
-        }
-    }
+post-patch {
+        reinplace "s|@@MP_ARCH_FLAGS@@|${mp_arch_flags}|" ${worksrcpath}/../configure
 }
 
 default_variants    +sdl
